### ISSUE TITLE
use DocumentLinkProvider to limit update calls

### ISSUE THREAD

**Issue by rioj7** [moalamri/vscode-inline-fold#115]
Implement a `DocumentLinkProvider` and return an empty array before calling an update on the decorator.

VSC will limit the calls to the `DocumentLinkProvider` while the user is typing, no calls when the selection changes.

**Comment by moalamri** [comment]
> Implement a `DocumentLinkProvider` and return an empty array before calling an update on the decorator.
> 
> VSC will limit the calls to the `DocumentLinkProvider` while the user is typing, no calls when the selection changes.

Hey @rioj7 
May I know what this implementation is going to fix (is the extension causing a bug)? Or you are suggesting a different way of triggering the decorator? 
thanks

**Comment by rioj7** [comment]
@moalamri If I read the code correct you are now triggering on undo and redo on `OnDocumentSelectionChange`. If you do it with a `DocumentLinkProvider` it will happen on any change of content when the user has stopped typing. wait time about 1sec.

**Comment by moalamri** [comment]
Hey @rioj7 

Thank you for explaining, I have tested your suggestion and you are right it does limit the triggering and gets fired in a second. however, it also gets triggered after typing (as you have described) which will in fact add an extra trigger (because `onDidChangeTextEditorSelection` gets fired while typing), so if we were to replace both `onDidChangeTextEditorSelection` and `onDocumentSelectionChange` with `DocumentLinkProvider` the extension will apply the decoration late and it will lose the ability to immediately unfold on selection.
Another issue is lanuageid, which is not a big issue, but it added some extra steps to set the current languageid, and if I were to use a wildcard `*` as languageid it will for some reason trigger the event 4 times when a workspace is newly opened.

To add some info, the whole point of using the events limiter beside performance is to overcome this [bug](https://github.com/microsoft/vscode/issues/154977).

Btw, thanks for the API, I have used it in a private extension at work 👍🏻

**Comment by rioj7** [comment]
@moalamri I only wanted to make a suggestion and did not examine the whole code and the consequences. Before I found the `DocumentLinkProvider` I used the `SelectionChanged` event and that triggered my code also when there was no change in the text. And the update on the screen (WebView) lagged behind when you typed because it got called on every key press.

I was looking how you did the actual folding. And found that you use the same trick with the `textDecoration` that I found in the **Power Mode** extension, used it in **Better Whitespace V2**. After creating an issue to get more CSS properties in the decorator.
